chain/websocketd.py

```diff
@@ -23,29 +23,96 @@
+tag_zmq_sock = {}
+tag_subscribers = {}
 
 @websockets.route('/<tag>')
 def site_socket(ws, tag):
     logger.info('ws client connected for tag "%s"' % tag)
-    zmq_sock = zmq_ctx.socket(zmq.SUB)
-    zmq_sock.connect(ZMQ_PASSTHROUGH_URL_PUB)
-    # note that flask gives us tag as a unicode string
-    zmq_sock.setsockopt_string(zmq.SUBSCRIBE, tag)
+    if tag in tag_zmq_sock:
+        tag_subscribers[tag].add(ws)
+    else:
+        zmq_sock = zmq_ctx.socket(zmq.SUB)
+        zmq_sock.connect(ZMQ_PASSTHROUGH_URL_PUB)
+        # note that flask gives us tag as a unicode string
+        zmq_sock.setsockopt_string(zmq.SUBSCRIBE, tag)
+        zmq_socks.add(zmq_sock)
+        zmq_socks_tag[zmq_sock] = tag
+        tag_zmq_sock[tag] = zmq_sock
+        tag_subscribers[tag] = {ws}
     while True:
-        in_data = zmq_sock.recv()
-        # in case we later subscribe to multiple topic, get the topic from the
-        # incoming message
-        msg_tag, _, msg = in_data.partition(' ')
-        logger.debug('Received on tag "%s": %s' % (msg_tag, msg))
         try:
-            ws.send(msg)
+            # Read from ws socket continuously.
+            # XXX:  This builds up a lot of threads.  Is there ANY way we could do
+            #       this for many connected ws clients in a non-blocking fashion (like select).
+            # NOTE that ws dies if site_socket function exits...
+            received = ws.receive()
+            logger.info("Received from ws client: %s" % received)
         except Exception as e:
-            logger.info('Caught Error sending to client: %s' % e)
+            tag_subscribers[tag].remove(ws)
+            if len(tag_subscribers) == 0:
+                close_socket(tag_zmq_sock[tag])
             break
-    logger.info('Disconnecting ZMQ Socket for tag "%s"' % tag)
-    zmq_sock.disconnect(ZMQ_PASSTHROUGH_URL_PUB)
     # return an empty response so Flask doesn't complain
     return ''
 
+def close_socket(zmq_sock):
+    tag = zmq_socks_tag[zmq_sock]
+    logger.info('Disconnecting ZMQ Socket for tag "%s"' % tag)
+    try:
+        zmq_sock.disconnect(ZMQ_PASSTHROUGH_URL_PUB)
+    except Exception as e:
+        logger.info(str(e))
+    zmq_socks.remove(zmq_sock)
+    del zmq_socks_tag[zmq_sock]
+    del tag_zmq_sock[tag]
+    subscribers = tag_subscribers[tag]
+    del tag_subscribers[tag]
+    for ws in tag_subscribers:
+        try:
+            ws.close()
+        except Exception:
+            logger.info('Could not close ws socket cleanly.')
+
+def select_zmq_socks():
+    logger.info("Starting select loop over ZMQ sockets")
+    while True:
+        gevent.sleep(seconds=0.0625, ref=True)
+        while len(zmq_socks) == 0:
+            gevent.sleep(seconds=0.0625, ref=True)
+        # It's important to have a timeout for the select loop, because
+        #   if more sockets are added, they won't be slected for until
+        #   the next iteration of the while loop
+        socks_list = list(zmq_socks)
```

Comment:
another way to do this without the quick-spin loop would be to add an internal ZMQ socket called `select_wake` or something and add it to the list of sockets we're selecting. Then whenever you wanted to kick the select process you could just write a byte to that socket and it would wake up and pick up any new sockets.

Not a blocking issue though.